solr/core/src/test/org/apache/solr/search/TestBackgroundJoinWarmer.java

```diff
@@ -0,0 +1,183 @@
+
+    }
+
+
+    public static String update(SolrCore core, String xml) throws Exception {
+        return update(core, xml, null);
+    }
+
+    public static String update(SolrCore core, String xml, SolrParams params) throws Exception {
+        DirectSolrConnection connection = new DirectSolrConnection(core);
+        SolrRequestHandler handler = core.getRequestHandler("/update");
+        return connection.request(handler, params, xml);
+    }
+
+    @Test
+    public void testScoreJoin() throws Exception {
+        doTestJoin("{!join "+(random().nextBoolean() ? "score=none" : ""));
+    }
+
+    void doTestJoin(String joinPrefix) throws Exception {
+        long toSearcherOpenTime = getSearcherOpenTime();
+        assertUserCacheStats("fromCore", "size", 0);
+        assertTrue(toSearcherOpenTime>0L);
+        for(long hits=0L; hits<2L; hits++) {
+            assertJQ(req(qOrFq(joinPrefix + " from=dept_id_s to=dept_s cacheEventually=true fromIndex=fromCore}cat:dev", "fl", "id",
+                    "debugQuery", random().nextBoolean() ? "true" : "false"))
+                    , "/response=={'numFound':3,'start':0,'numFoundExact':true,'docs':[{'id':'1'},{'id':'4'},{'id':'5'}]}"
+            );
+            // assert user cache hit
+            assertUserCacheStats("fromCore", "size", 1,
+                    "hits", hits, "cumulative_hits", hits, //0, 1
+                    "inserts", 1L , "cumulative_inserts", 1L);
+        }
+        update(fromCore, delQ("id:10"));
+        update(fromCore, add(doc("id", "10", "id_s_dv", "10", "dept_id_s", "Marketing",
+                "text", "now they sell", "cat", "dev")));
+        update(fromCore, commit());
+        // change in fromCore isn't visible due to user cache cache
+        assertJQ(req(qOrFq( joinPrefix + " from=dept_id_s to=dept_s cacheEventually=true fromIndex=fromCore}cat:dev", "fl", "id",
+                "debugQuery", random().nextBoolean() ? "true" : "false"))
+                , "/response=={'numFound':3,'start':0,'numFoundExact':true,'docs':[{'id':'1'},{'id':'4'},{'id':'5'}]}"
+        );
+        // nothing have changed ...
+        assertUserCacheStats("fromCore", "size", 1,
+                "hits", 2L, "cumulative_hits", 2L, // inc from the last check
+                "inserts", 1L , "cumulative_inserts", 1L);
+
+        if (true) {
+            update(fromCore, commit(), new MapSolrParams(Map.of("post-processor", "refresh-join-caches")));
+            assertUserCacheStats("fromCore", "size", 1,
+                    "hits", 0L,
+                    "cumulative_hits", 0L, // that's because it warms itself
+                    "inserts", 0L ,
+                    "cumulative_inserts", 0L); // that's because it warms itself
+        } else { // commit into main index refreshes cache as well
+            assertU(add(doc("id", "99999999")));
+            assertU(commit()); // TODO this causes an error.
+            // cached join leaks as query result key,
```

Comment:
This is a problem so far. Doc set from old searcher breaks warming of fresh searcher. Have no solution.